python/src/main/python/drivers/runspv.py

```diff
@@ -0,0 +1,944 @@
+    elif shader_basename.endswith('.asm'):
+        result += remove_end(shader_basename, '.asm') + '.spv'
+        cmd = spirvas_path() + ' ' + shader + ' -o ' + result
+        subprocess.check_call(cmd, shell=True, timeout=TIMEOUT_RUN)
+    elif shader_basename.endswith('.spv'):
+        result += shader_basename
+        shutil.copy(shader, result)
+    else:
+        assert False, 'unexpected shader extension: {}'.format(shader)
+
+    return result
+
+
+################################################################################
+# Android general
+
+
+ANDROID_SDCARD_GRAPHICSFUZZ_DIR = '/sdcard/graphicsfuzz'
+# Amber cannot be made executable under /sdcard/, hence we work under /data/local/tmp
+ANDROID_DEVICE_DIR = '/data/local/tmp'
+ANDROID_DEVICE_GRAPHICSFUZZ_DIR = ANDROID_DEVICE_DIR + '/graphicsfuzz'
+ANDROID_AMBER_NDK = 'amber_ndk'
+ANDROID_LEGACY_APP = 'com.graphicsfuzz.vkworker'
+TIMEOUT_APP = 30
+
+
+def adb_helper(adb_args, check, stdout: Union[None, int, IO[Any]]):
+    adb_cmd = 'adb ' + adb_args
+
+    try:
+        p = subprocess.run(
+            adb_cmd,
+            shell=True,
+            check=check,
+            timeout=TIMEOUT_RUN,
+            stdout=stdout,
+            universal_newlines=True,
+        )
+        return p
+
+    except subprocess.TimeoutExpired as err:
+        print('ERROR: adb command timed out: ' + err.cmd)
+        raise err
+
+
+def adb_check(adb_args, stdout: Union[None, int, IO[Any]] = subprocess.PIPE):
+    return adb_helper(adb_args, True, stdout)
+
+
+def adb_can_fail(adb_args, stdout: Union[None, int, IO[Any]] = subprocess.PIPE):
+    return adb_helper(adb_args, False, stdout)
+
+
+def stay_awake_warning():
+    res = adb_check('shell settings get global stay_on_while_plugged_in')
+    if str(res.stdout).strip() == '0':
+        print('\nWARNING: please enable "Stay Awake" from developer settings\n')
+
+
+def is_screen_off_or_locked():
+    """
+    :return: True: the screen is off or locked. False: unknown.
+    """
+    res = adb_can_fail('shell dumpsys nfc')
+    if res.returncode != 0:
+        return False
+
+    stdout = str(res.stdout)
+    # You will often find "mScreenState=OFF_LOCKED", but this catches OFF too, which is good.
+    if stdout.find('mScreenState=OFF') >= 0:
+        return True
+    if stdout.find('mScreenState=ON_LOCKED') >= 0:
+        return True
+
+    return False
+
+
+def prepare_device(wait_for_screen: bool, using_legacy_worker: bool):
+    adb_check('logcat -c')
+
+    if using_legacy_worker:
+        # If the legacy worker is being used, give it the right permissions, stop it if already running, and get its
+        # working directory ready.
+        adb_check('shell pm grant com.graphicsfuzz.vkworker android.permission.READ_EXTERNAL_STORAGE')
+        adb_check('shell pm grant com.graphicsfuzz.vkworker android.permission.WRITE_EXTERNAL_STORAGE')
+        adb_can_fail('shell am force-stop ' + ANDROID_LEGACY_APP)
```

Comment:
Force-stopping the worker everytime is good to make sure it is starting in a clean state, but I just want to point out that it is not strictly required and is likely to have performance impacts.

Reply:
I would vote to keep this because this is how it was before and it is our legacy worker anyway.